### PR TITLE
[WIP] Replace Zygote adjoints with ChainRules' rrules

### PR DESCRIPTION
**Pull request by sharanry**
https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/issues/116#issuecomment-671212997

Currently a lot of AD tests fail. Still need to work on revamping the AD tests to use [ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl).

TODO

- [ ] Use `rrule_test` for each of the `rrules`. Plain usage of them results in errors with `FiniteDifferences`. Figure of if this is due to the incorrect rrules.
- [ ] `test/utils.jl` seems pass with minimal imports but fail other wise. Figure out why.

### CODE REVIEW

File: src/chainrules.jl [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    back(Δ::AbstractMatrix) = (NO_FIELDS, Δ,)
+    function back(Δ::AbstractVector{<:AbstractVector{<:Real}})
+        throw(error("In slow method"))
+    end
+    return RowVecs(X), back
+end
+
+
+# rules for transforms
+function ChainRulesCore.rrule(::typeof(Base.map), t::Transform, X::ColVecs)
```
Comment (devmotion) [inline src/chainrules.jl:84]: Unfortunately we can't move these rules for `map` from ZygoteRules to ChainRules. https://github.com/JuliaDiff/ChainRulesCore.jl/issues/68 has to be fixed first.

File: src/chainrules.jl [hunk 2]
```diff
@@ -0,0 +1,90 @@
+## rules for Delta
+function ChainRulesCore.rrule(::typeof(evaluate), s::Delta, x::AbstractVector, y::AbstractVector)
+  evaluate(s, x, y), Δ -> begin
```
Comment (devmotion) [inline src/chainrules.jl:3]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    (NO_FIELDS, DoesNotExist(), DoesNotExist(), DoesNotExist())
+  end
+end
+
+function ChainRulesCore.rrule(::typeof(Distances.pairwise), d::Delta, X::AbstractMatrix, Y::AbstractMatrix; dims=2)
+  D = Distances.pairwise(d, X, Y; dims = dims)
+  if dims == 1
+      return D, Δ -> (NO_FIELDS, DoesNotExist(), DoesNotExist(), DoesNotExist())
```
Comment (devmotion) [inline src/chainrules.jl:11]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  else
+      return D, Δ -> (NO_FIELDS, DoesNotExist(), DoesNotExist(), DoesNotExist())
```
Comment (devmotion) [inline src/chainrules.jl:13]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  end
+end
+
+function ChainRulesCore.rrule(::typeof(Distances.pairwise), d::Delta, X::AbstractMatrix; dims=2)
+  D = Distances.pairwise(d, X; dims = dims)
+  if dims == 1
+      return D, Δ -> (NO_FIELDS, DoesNotExist(), DoesNotExist())
```
Comment (devmotion) [inline src/chainrules.jl:20]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  else
+      return D, Δ -> (NO_FIELDS, DoesNotExist(), DoesNotExist())
```
Comment (devmotion) [inline src/chainrules.jl:22]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl [hunk 7]
```diff
@@ -0,0 +1,90 @@
+## rules for Delta
+function ChainRulesCore.rrule(::typeof(evaluate), s::Delta, x::AbstractVector, y::AbstractVector)
+  evaluate(s, x, y), Δ -> begin
+    (NO_FIELDS, DoesNotExist(), DoesNotExist(), DoesNotExist())
+  end
+end
+
+function ChainRulesCore.rrule(::typeof(Distances.pairwise), d::Delta, X::AbstractMatrix, Y::AbstractMatrix; dims=2)
+  D = Distances.pairwise(d, X, Y; dims = dims)
+  if dims == 1
```
Comment (devmotion) [inline src/chainrules.jl:10]: It seems the check is not needed here.

File: src/chainrules.jl [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+      return D, Δ -> (NO_FIELDS, DoesNotExist(), DoesNotExist(), DoesNotExist())
+  else
+      return D, Δ -> (NO_FIELDS, DoesNotExist(), DoesNotExist(), DoesNotExist())
+  end
+end
+
+function ChainRulesCore.rrule(::typeof(Distances.pairwise), d::Delta, X::AbstractMatrix; dims=2)
+  D = Distances.pairwise(d, X; dims = dims)
+  if dims == 1
```
Comment (devmotion) [inline src/chainrules.jl:19]: The check is not needed here.

File: src/chainrules.jl [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      return D, Δ -> (NO_FIELDS, DoesNotExist(), DoesNotExist())
+  else
+      return D, Δ -> (NO_FIELDS, DoesNotExist(), DoesNotExist())
+  end
+end
+
+## rules for DotProduct
+function ChainRulesCore.rrule(::typeof(evaluate), s::DotProduct, x::AbstractVector, y::AbstractVector)
+  dot(x, y), Δ -> begin
```
Comment (devmotion) [inline src/chainrules.jl:28]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    (NO_FIELDS, nothing, Δ .* y, Δ .* x)
```
Comment (devmotion) [inline src/chainrules.jl:29]: Maybe use `@thunk`?

File: src/chainrules.jl [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+  end
+end
+
+function ChainRulesCore.rrule(::typeof(Distances.pairwise), d::DotProduct, X::AbstractMatrix, Y::AbstractMatrix; dims=2)
+  D = Distances.pairwise(d, X, Y; dims = dims)
+  if dims == 1
+      return D, Δ -> (NO_FIELDS, nothing, Δ * Y, (X' * Δ)')
```
Comment (devmotion) [inline src/chainrules.jl:36]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  else
+      return D, Δ -> (NO_FIELDS, nothing, (Δ * Y')', X * Δ)
```
Comment (devmotion) [inline src/chainrules.jl:38]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl (hunk 11 above)
Comment (devmotion) [reply src/chainrules.jl:36]: Maybe use thunks?

File: src/chainrules.jl (hunk 12 above)
Comment (devmotion) [reply src/chainrules.jl:38]: Maybe use thunks?

File: src/chainrules.jl [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+  end
+end
+
+function ChainRulesCore.rrule(::typeof(Distances.pairwise), d::DotProduct, X::AbstractMatrix; dims=2)
+  D = Distances.pairwise(d, X; dims = dims)
+  if dims == 1
+      return D, Δ -> (NO_FIELDS, nothing, 2 * Δ * X)
```
Comment (devmotion) [inline src/chainrules.jl:45]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+  else
+      return D, Δ -> (NO_FIELDS, nothing, 2 * X * Δ)
```
Comment (devmotion) [inline src/chainrules.jl:47]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+  end
+end
+
+## rules for Sinus
+function ChainRulesCore.rrule(::typeof(evaluate), s::Sinus, x::AbstractVector, y::AbstractVector)
+  d = (x - y)
+  sind = sinpi.(d)
+  val = sum(abs2, sind ./ s.r)
+  gradx = 2π .* cospi.(d) .* sind ./ (s.r .^ 2)
+  val, Δ -> begin
```
Comment (devmotion) [inline src/chainrules.jl:57]: Pullback should be defined as a properly named function (see ChainRules docs).

File: src/chainrules.jl [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    (NO_FIELDS, (r = -2Δ .* abs2.(sind) ./ s.r,), Δ * gradx, - Δ * gradx)
```
Comment (devmotion) [inline src/chainrules.jl:58]: Use thunks?

File: src/chainrules.jl [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+function vecs_pullback(Δ::AbstractVector{<:AbstractVector{<:Real}})
+    throw(error("In slow method"))
+end
+function ChainRulesCore.rrule(::typeof(ColVecs), X::AbstractMatrix)
+    return ColVecs(X), vecs_pullback
+end
+
+function ChainRulesCore.rrule(::typeof(RowVecs), X::AbstractMatrix)
+    return RowVecs(X), vecs_pullback
+end
```
Comment (sharanry) [inline src/chainrules.jl:80]: @willtebbutt How do we go about defining  `rrule`s for constructors? 

For example, here `typeof(ColVecs)` results in `UnionAll` which is probably not what we want.

Comment (sharanry) [reply src/chainrules.jl:80]: Would it make sense to replace `typeof(ColVecs)` with `Type{ColVecs}`?

**Comment by devmotion** [comment]
Maybe https://github.com/JuliaDiff/ChainRulesCore.jl/issues/150 could simplify some definitions?

File: src/chainrules.jl [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+
+
+# rules for ColVecs and RowVecs
+vecs_pullback(Δ::NamedTuple) = (NO_FIELDS, Δ.X,)
+vecs_pullback(Δ::AbstractMatrix) = (NO_FIELDS, Δ,)
+function vecs_pullback(Δ::AbstractVector{<:AbstractVector{<:Real}})
+    throw(error("In slow method"))
+end
+
+function ChainRulesCore.rrule(::Type{ColVecs}, X::AbstractMatrix)
```
Comment (willtebbutt) [inline src/chainrules.jl:75]: I think the right thing to do here would be `::Type{<:ColVecs}`, would it not?

File: src/chainrules.jl [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    return ColVecs(X), vecs_pullback
+end
+
+function ChainRulesCore.rrule(::Type{RowVecs}, X::AbstractMatrix)
```
Comment (willtebbutt) [inline src/chainrules.jl:79]: Same here?

**Review by willtebbutt — COMMENTED** [review]
This is looking good. Just needs to use the tests in `ChainRulesCore` -- you might need to define `to_vec` for a few things, but that's about it.

**Comment by theogf** [comment]
Replaced by #208

**Pull request closed without merging**